gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -258,37 +260,53 @@ protected void addSingleTaskWriterOutputToExistingDir(Path writerOutputDir, Path
         continue;
       }
       String pathSuffix = taskOutputFile
-          .substring(taskOutputFile.indexOf(writerOutputDir.toString()) + writerOutputDir.toString().length() + 1);
+              .substring(taskOutputFile.indexOf(writerOutputDir.toString()) + writerOutputDir.toString().length() + 1);
       Path publisherOutputPath = new Path(publisherOutputDir, pathSuffix);
       WriterUtils.mkdirsWithRecursivePermission(this.publisherFileSystemByBranches.get(branchId),
           publisherOutputPath.getParent(), this.permissions.get(branchId));
 
       LOG.info(String.format("Moving %s to %s", taskOutputFile, publisherOutputPath));
       parallelRunner.movePath(taskOutputPath, this.publisherFileSystemByBranches.get(branchId),
-          publisherOutputPath, Optional.<String> absent());
+              publisherOutputPath, Optional.<String> absent());
     }
   }
 
+  private Path targetPath(Path dstDir, Path src, Path path) {
+    String pathStr = path.toString();
+    String srcStr = src.toString();
+    String relative = pathStr.substring(pathStr.indexOf(srcStr) + srcStr.length() + 1);
+    return new Path(dstDir, relative);
+  }
+
+
   protected void addWriterOutputToExistingDir(Path writerOutputDir, Path publisherOutputDir,
       WorkUnitState workUnitState, int branchId, ParallelRunner parallelRunner) throws IOException {
     boolean preserveFileName = workUnitState.getPropAsBoolean(ForkOperatorUtils.getPropertyNameForBranch(
         ConfigurationKeys.SOURCE_FILEBASED_PRESERVE_FILE_NAME, this.numBranches, branchId), false);
 
-    // Go through each file in writerOutputDir and move it into publisherOutputDir
-    for (FileStatus status : this.writerFileSystemByBranches.get(branchId).listStatus(writerOutputDir)) {
-
-      // Preserve the file name if configured, use specified name otherwise
-      Path finalOutputPath =
-          preserveFileName
-              ? new Path(publisherOutputDir,
-                  workUnitState.getProp(ForkOperatorUtils.getPropertyNameForBranch(
-                      ConfigurationKeys.DATA_PUBLISHER_FINAL_NAME, this.numBranches, branchId)))
-          : new Path(publisherOutputDir, status.getPath().getName());
+    FileSystem fs = this.writerFileSystemByBranches.get(branchId);
+    RemoteIterator<LocatedFileStatus> iterator = fs.listFiles(writerOutputDir, true);
```

Comment:
Is it because you want to do the recursive list? The main issue with your approach is that it breaks atomicity of publishing. The current implementation tries to publish at the directory level -- a single rename publishes all files in a single directory. This is very important to avoid race conditions where downstream consumers may see partial data.

I think I understand the underlying issue but the fix may be slightly trickier. Before adding the partitioned writer the publishing happened at the level of directories which are direct children of the writer output directory. With the partitioned writer, that may not be the case any more and the atomic publish needs to happen at a deeper level but I don't think doing it at file level is the right thing to do in most cases.